### DISCUSSION TITLE
Google Summer of Code (GSOC) 2023

### DISCUSSION THREAD

**Discussion by unkcpz** [aiidateam/aiida-core#5902] [Announcements]
AiiDA is back on track in [Google Summer of Code (GSOC) 2023](https://summerofcode.withgoogle.com/) as the sub-organization under the [NumFocus](https://numfocus.org/) umbrella!

This thread is for any discussions leading up to GSOC 2023. If you're interested in working on AiiDA as a student, use this thread to say hi and ask any questions you may have.
Have a look at our [project ideas](https://github.com/aiidateam/aiida-core/wiki/GSoC-2023-Projects) for GSOC 2023, and see the [NumFOCUS guide for GSOC students](https://github.com/numfocus/gsoc/blob/master/CONTRIBUTING-students.md).

Just as a reminder, here is an excerpt from the [GSOC 2023 timeline](https://developers.google.com/open-source/gsoc/timeline):

February 22 - 18:00 UTC
- List of accepted mentoring organizations published

February 22 - March 19
- Potential GSoC contributors discuss application ideas with mentoring organizations

March 20 - 18:00 UTC
- GSoC contributor application period begins

April 4 - 18:00 UTC
- GSoC contributor application deadline

April 27 - 18:00 UTC
- GSoC contributor proposal rankings due from Org Admins

May 4 - 18:00 UTC
- Accepted GSoC contributor projects announced

The GSOC 2023 mentors

**Comment by AhmedBasem20** [comment]
Hello all 👋
I'm Ahmed, I'm interested in working on AiiDA, especially on the ranking system for aiida registry project.
Also, I'm exploring the `aiida-core` codebase for the last two weeks. I'm learning a lot from that.

Thank you!

**Reply by unkcpz** [reply]
Hi Ahmed,
Glad you are interested in this and I see you already got a PR merged, very nice!
We (thanks to @ltalirz who initially bring up the idea) update the details of the "ranking system" task. You may want to have a look again and see if everything is more clear. 
Best,
Jason

**Reply by AhmedBasem20** [reply]
Hi Jason,
Thanks for updating the details. I can see there are some interesting features presented as well, sounds great!
At this time, I'm planning to write an initial GSOC proposal for the ranking system project and also continue contributing to `aiida-core`.
If I miss something here or if you have additional tips please let me know.

**Comment by AhmedBasem20** [comment]
Hi, I'm setting up the development environment on my local windows machine.
I used conda virtual environment and installed the required dependencies,
but there are some problems that appear when I try to run or test any file. I searched those errors but I couldn't figure out where is the problem.

For example when running `python aiida-core\aiida\cmdline\commands\cmd_node.py`

I got the following errors:
```
Traceback (most recent call last):
  File "D:\aiida-vm\aiida-core\aiida\manage\configuration\config.py", line 91, in from_file
    with open(filepath, 'rb') as handle:
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\Infotech\\.aiida\\config.json'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\aiida-core\aiida\cmdline\commands\cmd_node.py", line 17, in <module>
    from aiida.cmdline.commands.cmd_verdi import verdi
  .
  .
  .
  .
  File "D:\aiida-vm\aiida-core\aiida\manage\configuration\config.py", line 544, in _atomic_write
    os.rename(handle.name, self.filepath)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\Infotech\\.aiida\\tmpdwkvt6fk' -> 'C:\\Users\\Infotech\\.aiida\\config.json'
```

**Reply by unkcpz** [reply]
Hi @AhmedBasem20, it seems you are using Windows, so you need Windows Subsystem for Linux (WSL) to run AiiDA, please find detail from the documentation https://aiida.readthedocs.io/projects/aiida-core/en/latest/intro/install_system.html#intro-get-started-system-wide-install

**Reply by AhmedBasem20** [reply]
Hi Jason, Thanks for answering. I'll check the documentation out.

**Comment by kunal00000** [comment]
Hey Everyone, 
I am Kunal, found the projects listed by AiiDA aligned with my technical knowlegde (Project 2 - Ranking system for AiiDA plugin registry). 
I would wish to contribute to AiiDA, I am proficient with HTML, CSS, JavaScript, Node js, Express, learning React Js.
what do you say of switching custom-built static HTML website to Next js website. It can help improve the performance and scalability of AiiDA-registry.
Let me know initial steps to contribute.

**Reply by ltalirz** [reply]
Cheers @kunal00000 !

There are other components in the AiiDA / Materials Cloud ecosystem written (or being written) in React, so for the long term it might be advantageous to stick with this framework unless there are good arguments against it.

So far, the registry web page is rather low-data, i.e. performance and scalability are not a concern. Making it more useful to users, however, is!

There are different aspects to this project (including better data collection from plugins), but a first step could just be to roughly reproduce the functionality of the current web page within React.

The web page is currently built in two steps, as you can see on the continuous integration tests:
https://github.com/aiidateam/aiida-registry/blob/41d56a5467572bdef5b65f8c03b0e28e6eb78199/.github/workflows/ci.yml#L40-L44
First the metadata is fetched into a JSON file,
then the web page is built from that file.

A first step could be to keep the metadata collection step frozen, and to replace the `aiida-registry html` step with React (keeping in mind that we want the order of plugins to be flexible & may want to introduce search functionality).

**Reply by kunal00000** [reply]
Sounds Good Leopold, 
Can I start making React web page with current functionalities(like prototype) before GSoC contribution period starts or starting proposal would be better?

**Reply by unkcpz** [reply]
Hi @kunal00000, 
thanks a lot for your interest. I'd suggest you take some time to get familiar with the AiiDA ecosystem. As we mentioned on the idea page, to be considered as a GSoC student, we ask you to make a small pull request to aiida-core or the repositories in our organization [aiidateam](https://github.com/aiidateam) and [aiidalab-team](https://github.com/aiidateam) - could be a simple bug fix, improving the documentation. We have a list of issues that are ["good first"](https://github.com/aiidateam/aiida-core/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) and proper for beginner contributors.  
As a mentor, I'd prefer to stick to the timeline and start the official mentoring after the student is selected. If you have any questions, post them here and I am happy to help.

Cheers,
Jason

**Reply by ltalirz** [reply]
@kunal00000 as you may have noticed, web dev is not our primary skill here - and we are always happy to learn.
After looking a bit into next.js, I understand why you suggested it, and it does look like a perfectly viable solution vs bare react, maybe better.
Happy to discuss should you submit a proposal.

**Comment by MikeW1121** [comment]
Hello everyone!
My name is Ziming Wang and I found project 1 - Easy Computer/Code setup very interesting and would love to contribute to it over the summer through GSOC. I have experience in Python HTML, and Javascript, and am currently learning more about the Jinja templating engine. I hope to gain experience working on impactful projects and meet more like-minded individuals. I can't wait to start contributing to AiiDA now and throughout the summer.

**Reply by unkcpz** [reply]
Hi Ziming, Glad you are interested in it. For this task, it is very related to the CLI system of AiiDA where we use [`click`](https://click.palletsprojects.com/en/8.1.x/) under the hood. There are some issues in `aiida-core` related to it, you can find from https://github.com/aiidateam/aiida-core/issues?q=is%3Aissue+is%3Aopen+CLI. 
Meanwhile, there are two [PRs](https://github.com/aiidateam/aiida-code-registry/pulls) in `aiida-code-registry` as the proof of concept for this task, you may get some inspiration on how to process it.
Best, Jason

**Reply by MikeW1121** [reply]
Thank you so much! I will definitely take a look at these resources.

**Comment by AhmedBasem20** [comment]
Hello @ltalirz,
Regarding the "ranking system" project, could you please elaborate more what is the additional information about calculations, data types and workflows to be extracted besides the entry point name?
Also, what do you think of using external APIs from PyPI or Github to get information about download times and repository stars?
Thanks!

**Reply by ltalirz** [reply]
Hi @AhmedBasem20 , sure!

The additional metadata I'm referring to is obtained by inspecting the `ProcessSpec` of the respective CalcJob/WorkChain plugins, just like how the `verdi plugin list` command works:

As you can see below, this information is quite detailed and can give you a good idea of the capabilities of the Calcjob/WorChain.
At the moment, all of this information is "dormant" inside the plugin, and we want to extract it into the registry.

```
$ verdi plugin list aiida.workflows quantumespresso.pw.relax
Description:

    Workchain to relax a structure using Quantum ESPRESSO pw.x.

Inputs:
                           base  Data           Inputs for the `PwBaseWorkChain` for the main relax loop.
                      structure  StructureData  The inputs structure.
                 base_final_scf  Data           Inputs for the `PwBaseWorkChain` for the final scf.
                  clean_workdir  Bool           If `True`, work directories of all called calculation will be cleaned at
                                                the end of execution.
max_meta_convergence_iterations  Int            The maximum number of variable cell relax iterations in the meta
                                                convergence cycle.
               meta_convergence  Bool           If `True` the workchain will perform a meta-convergence on the cell volume.
                       metadata
             volume_convergence  Float          The volume difference threshold between two consecutive meta convergence
                                                iterations.

Required inputs are displayed in bold red.

Outputs:
        output_parameters  Dict             The `output_parameters` output node of the successful calculation.
            remote_folder  RemoteData       Input files necessary to run the process will be stored in this folder
                                            node.
                retrieved  FolderData       Files that are retrieved by the daemon will be stored in this node. By
                                            default the stdout and stderr of the scheduler will be added, but one can
                                            add more by specifying them in `CalcInfo.retrieve_list`.
automatic_parallelization  Dict             The results of the automatic parallelization analysis if performed.
output_atomic_occupations  Dict
              output_band  BandsData        The `output_band` output node of the successful calculation if present.
           output_kpoints  KpointsData
         output_structure  StructureData    The successfully relaxed structure.
        output_trajectory  TrajectoryData
             remote_stash  RemoteStashData  Contents of the `stash.source_list` option are stored in this remote folder
                                            after job completion.

Required outputs are displayed in bold red.

Exit codes:

  0  The process finished successfully.
  1  The process has failed with an unspecified error.
  2  The process failed with legacy failure mode.
 10  The process returned an invalid output.
 11  The process did not register a required output.
401  the relax PwBaseWorkChain sub process failed
402  the final scf PwBaseWorkChain sub process failed

Exit codes that invalidate the cache are marked in bold red.


$ verdi plugin list aiida.calculations quantumespresso.pw
Description:

    `CalcJob` implementation for the pw.x code of Quantum ESPRESSO.

Inputs:
        kpoints  KpointsData               kpoint mesh or kpoint path
     parameters  Dict                      The input parameters that are to be used to construct the input file.
        pseudos  UpfData, UpfData          A mapping of `UpfData` nodes onto the kind name to which they should apply.
      structure  StructureData             The input structure.
           code  AbstractCode, NoneType    The `Code` to use for this job. This input is required, unless the
                                           `remote_folder` input is specified, which means an existing job is being
                                           imported and no code will actually be run.
   hubbard_file  SinglefileData, NoneType  SinglefileData node containing the output Hubbard parameters from a
                                           HpCalculation
       metadata
       monitors  Dict                      Add monitoring functions that can inspect output files while the job is
                                           running and decide to prematurely terminate the job.
parallelization  Dict, NoneType            Parallelization options. The following flags are allowed: npool  : The
                                           number of 'pools', each taking care of a group of k-points. nband  : The
                                           number of 'band groups', each taking care of a group of Kohn-Sham orbitals.
                                           ntg    : The number of 'task groups' across which the FFT planes are
                                           distributed. ndiag  : The number of 'linear algebra groups' used when
                                           parallelizing the subspace diagonalization / iterative orthonormalization.
                                           By default, no parameter is passed to Quantum ESPRESSO, meaning it will use
                                           its default.
  parent_folder  RemoteData, NoneType      An optional working directory of a previously completed calculation to
                                           restart from.
  remote_folder  RemoteData, NoneType      Remote directory containing the results of an already completed calculation
                                           job without AiiDA. The inputs should be passed to the `CalcJob` as normal
                                           but instead of launching the actual job, the engine will recreate the input
                                           files and then proceed straight to the retrieve step where the files of
                                           this `RemoteData` will be retrieved as if it had been actually launched
                                           through AiiDA. If a parser is defined in the inputs, the results are parsed
                                           and attached as output nodes as usual.
       settings  Dict, NoneType            Optional parameters to affect the way the calculation job and the parsing
                                           are performed.
      vdw_table  SinglefileData, NoneType  Optional van der Waals table contained in a `SinglefileData`.

Required inputs are displayed in bold red.

Outputs:
        output_parameters  Dict             The `output_parameters` output node of the successful calculation.
            remote_folder  RemoteData       Input files necessary to run the process will be stored in this folder
                                            node.
                retrieved  FolderData       Files that are retrieved by the daemon will be stored in this node. By
                                            default the stdout and stderr of the scheduler will be added, but one can
                                            add more by specifying them in `CalcInfo.retrieve_list`.
output_atomic_occupations  Dict
              output_band  BandsData        The `output_band` output node of the successful calculation if present.
           output_kpoints  KpointsData
         output_structure  StructureData    The `output_structure` output node of the successful calculation if
                                            present.
        output_trajectory  TrajectoryData
             remote_stash  RemoteStashData  Contents of the `stash.source_list` option are stored in this remote folder
                                            after job completion.

Required outputs are displayed in bold red.

Exit codes:

  0  The process finished successfully.
  1  The process has failed with an unspecified error.
  2  The process failed with legacy failure mode.
 10  The process returned an invalid output.
 11  The process did not register a required output.
100  The process did not have the required `retrieved` output.
110  The job ran out of memory.
120  The job ran out of walltime.
131  The specified account is invalid.
140  The node running the job failed.
150  {message}
301  The retrieved temporary folder could not be accessed.
302  The retrieved folder did not contain the required stdout output file.
303  The retrieved folder did not contain the required XML file.
304  The retrieved folder contained multiple XML files.
305  Both the stdout and XML output files could not be read or parsed.
310  The stdout output file could not be read.
311  The stdout output file could not be parsed.
312  The stdout output file was incomplete probably because the calculation got
     interrupted.
320  The XML output file could not be read.
321  The XML output file could not be parsed.
322  The XML output file has an unsupported format.
340  The calculation stopped prematurely because it ran out of walltime but the
     job was killed by the scheduler before the files were safely written to
     disk for a potential restart.
350  The parser raised an unexpected exception: {exception}
400  The calculation stopped prematurely because it ran out of walltime.
410  The electronic minimization cycle did not reach self-consistency.
461  The code failed with negative dexx in the exchange calculation.
462  The code failed during the cholesky factorization.
463  Too many bands failed to converge during the diagonalization.
481  The k-point parallelization "npools" is too high, some nodes have no
     k-points.
500  The ionic minimization cycle did not converge for the given thresholds.
501  Then ionic minimization cycle converged but the thresholds are exceeded in
     the final SCF.
502  The ionic minimization cycle did not converge after the maximum number of
     steps.
510  The electronic minimization cycle failed during an ionic minimization
     cycle.
511  The ionic minimization cycle converged, but electronic convergence was not
     reached in the final SCF.
520  The ionic minimization cycle terminated prematurely because of two
     consecutive failures in the BFGS algorithm.
521  The ionic minimization cycle terminated prematurely because of two
     consecutive failures in the BFGS algorithm and electronic convergence
     failed in the final SCF.
531  The electronic minimization cycle did not reach self-consistency.
541  The variable cell optimization broke the symmetry of the k-points.
710  The electronic minimization cycle did not reach self-consistency, but
     `scf_must_converge` is `False` and/or `electron_maxstep` is 0.

Exit codes that invalidate the cache are marked in bold red.
```

Extracting this information is a little more tricky since it requires the AiiDA plugin to be installed.

I've actually already extracted the [relevant code](https://github.com/aiidateam/aiida-registry/blob/master/bin/analyze_entrypoints.py) from AiiDA in the AiiDA registry;
the `aiida-registry test-install` command also worked locally for me (via docker).
and there is even already a CI workflow for it [here](https://github.com/aiidateam/aiida-registry/blob/4f62a17406dcb801bced4c179bd568d82ec83956/.github/workflows/ci.yml#L46-L80).

Unfortunately, for some reason I wasn't able to get docker to work properly in CI within the time available.

So, what remains to be done here, is to get it to work on CI, and then to integrate this information into the registry web site.

**Reply by ltalirz** [reply]
As for alternative ranking methods, the first indicator I would suggest to implement is the most recent release.

For this we will need to change the current logic of the registry to commit the JSON file that contains all the collected information to git (this will also have other benefits for collecting statistics of the registry over time).
From looking at the changes of the JSON over time, one can construct this indicator.

Other indicators like the ones you mention may not be available for all plugins, but could still be useful. We're open to creative ideas here

**Reply by AhmedBasem20** [reply]
Hi Leopold, Sorry for the delay.

You have already done most of the work on this task. As I'm new to docker and CI, I don't really understand what is the problem.
But I find that It's a good opportunity to learn more and give [that issue](https://github.com/aiidateam/aiida-registry/issues/215) a try.

And of course, Thanks a lot for the explanation!

**Reply by AhmedBasem20** [reply]
For the ranking method, I got your point of obtaining the recent release and other statistics by looking at changes in the JSON over time, but I am not sure I understand committing the JSON file to git part. should that happen in the CI workflow?

**Reply by ltalirz** [reply]
> but I am not sure I understand committing the JSON file to git part. should that happen in the CI workflow?

Yes, that will happen on CI.

We'll see how to best do this - we could either make sure the JSON file is included in the `gh-pages` branch that includes the built web application, or (if necessary) we could have a separate branch of the repo that just tracks the progress of the plugins.json file.

**Reply by AhmedBasem20** [reply]
Alright, I'll keep working on and researching those tasks. Thanks!

**Comment by Seavus999** [comment]
Dear mentors,

My name is Seavus Fan. I am master student of computer science at Georgia Tech. I saw the information about AiiDa in a gitter room. I am super passionate about web development. Also, I have sufficient time during summer and I would be much obliged that I could have the opportunity to work with experienced mentor from AiiDa throughout the summer. 

I have knowledge of python, HTML, CSS, Javascript and RESTful API. However, I am new to open source. I have looked through the three ideas and some resources provided by you in the discussion board, but currently I am not sure which project I am more interested in. I would try to aiida core during the next month. 

Thanks very much!

Best wishes,

Yilan

**Comment by BAIKANI-ANIL-YADAV** [comment]
Hello all,

I'm Anil Yadav, I'm interested in doing the second project called (Ranking system for AiiDA plugin registry) from the given three  projects of AiiDA. I love to contribute to AiiDA,Im proficient in required skills for this project. I'm so excited to do this project and I need to know more about this project.

Thanks!

**Comment by devanshcache** [comment]
Hi everyone,

I hope I am not too late. During my research, I found your organization and the project that I thought would be interesting to me and a good place to start. About me, I am a working professional currently working as a Full Stack Engineer where we are delivering High Accuracy Indoor Positioning system. I have experience with Go, Python, Reactjs, Java, JavaScript, and  DevOps tools. 

The open-source world is new to me, but I'd love to learn and contribute. I am interested in two projects which are "**Easy Computer/Code setup with AiiDA code registry**"  and  "**Explore the AiiDA node graph in the browser**".

Thank you very much!

**Comment by unkcpz** [comment]
Hi all,

Thanks a lot to all for saying hi and in particular to those of you who tried making contributions with "good first issues". 

The application of proposals for the projects is **OPEN** today. Please contact me and send me your draft proposal to me directly (jusong.yeu@gmail.com) if you need any preliminary feedback and discuss more project details. 
Note that the deadline for submitting the proposal is **_4th April_**.

Best,
Jason

**Comment by lemythe** [comment]
hi

**Comment by Mahhheshh** [comment]
Hello everyone 👋 , 
I’m interested in working on Project 3 - Explore the AiiDA node graph in the browser. 
Python and the Rest API are two things I've worked with before, and I'm eager to learn more about AiiDA. Would you kindly elaborate on how AIIDA stores the entities in its database so that they can later be used to create a directed graph?
Thank you!

**Reply by ltalirz** [reply]
Hey @Mahhheshh , thanks for the interest.

I would suggest to start reading here: https://aiida.readthedocs.io/projects/aiida-core/en/latest/internals/storage/architecture.html

You'll also find documentation about AiiDA's current REST API in the docs.

There is an existing implementation of the graph explorer (still written in AngularJS) https://github.com/materialscloud-org/aiida-explorer
that powers https://www.materialscloud.org/explore/menu

That should be enough to get you started; feel free to reach out to the mentors of the project with questions

**Reply by Mahhheshh** [reply]
Thanks, @ltalirz for providing the resources. I will check them out. 
Could you please tell me which REST API framework will be used for the project?

**Reply by ltalirz** [reply]
The REST API that currently ships with aiida-core is a somewhat homegrown flask application, using the sqlachemy ORM that is used in AiiDA.

There is a more canonical implementation based on pydantic+fastapi in https://github.com/aiidateam/aiida-restapi ; perhaps the contact persons @unkcpz and @superstar54  for this project could comment on which implementation they are intending to use

**Comment by ltalirz** [comment]
This is just a reminder that proposal submission will close April 4th.

We expect prospective students to explore the project ideas [1] and come up with a proposal draft independently, but we're always happy to answer specific questions and give you feedback on your proposals.

Thanks to those who already submitted - looking forward to a great GSOC!

[1] By the way, we're always happy to discuss ideas not listed there as well.

**Comment by kartikeysaran** [comment]
Hi everyone,
I am senior student at University of Delhi. I came across the project ``` Easy Computer/Code setup with AiiDA code registry ```  and found it interesting would love to work on that this summer. :)

**Comment by kartikeysaran** [comment]
Hi @ltalirz , I hope you're doing well. I would love to get your feedback on my project proposal. Could you please let me know the best way to reach out to you for a review? Thank you for your time!

**Reply by unkcpz** [reply]
Hi @kartikeysaran, you can share the proposal to my Gmail: jusong.yeu[at]gmail.com
This task was initially brought up by me, I am happy to have a look and give you feedback.

Best,
Jason

**Reply by kartikeysaran** [reply]
Hi @unkcpz, Thank you so much for offering to review my project proposal. I really appreciate your willingness to take the time to provide feedback and help me improve my work.  I will send an email to you :)

**Comment by 0xt3j4s** [comment]
Dear mentors,

I am Tejas Jamdade, a junior pursuing my Bachelor of Technology in Materials Engineering at IIT Jodhpur. I got to know about AiiDA through one of the professors at IIT Jodhpur, Dr Appala Naidu Gandi. I am interested in working with AiiDA, especially on the AiiDA code registry and the overhauling and revamping the AiiDA plugin registry website. I am proficient in python, C, C++, JavaScript, and HTML, and I have some experience in web development and REST API as well. Currently, I am learning React.
Also, I am working on a few issues in the `aiida-core` repository. 

I look forward to working on the exciting project opportunities that AiiDA has to offer.

--
Yours sincerely
Tejas Jamdade
Junior, IIT Jodhpur

**Comment by unkcpz** [comment]
Dear all,

Kindly remind that the deadline for proposal submission is on **Tue, 4th Apr,** I guess you can find the precise time and date in your time zone from your student dashboard.
For students who already send us your proposal, we gave the feedback so you can further polish and improve it. Please don't forget (and don't wait until last second) to submit the final version before the deadline :smile: 

Best wishes!
Jason

**Comment by epsilon02fft** [comment]
Hi 
My name is Nidhi. I am interested in working on Project 1. I would really like to learn and contribute to AiiDA.

Best Regards
Nidhi

**Reply by snagar4** [reply]
Hi Nidhi, 
Probably in the same boat as you right now, I think we're too late since the submission deadline is tomorrow.
You may reach out to me at snagar4@asu.edu and we can work on some projects. I have been in computational modelling for a while using DFT and MD. 
Best,
Sainyam

**Reply by ltalirz** [reply]
Cheers @snagar4 , thanks for the friendly words and helping make our community more welcoming.
It's much appreciated!

**Reply by epsilon02fft** [reply]
Hi @snagar4 
Thank you so much for your reply. I would really like to work on some projects. I will email you soon and keep contact.

Best,
Nidhi

**Comment by ltalirz** [comment]
Thank you everyone for your questions, comments, and proposals.
We'll be reviewing the proposals we've received, and by May 4th the winning proposals will be announced.

If you don't make it this year, don't worry - we'll be around for the next GSOC :-)

Best wishes,
Leopold
